servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultBlockingIterableProcessor.java

```diff
@@ -144,43 +120,33 @@ public boolean hasNext() {
                 return true;
             }
 
-            final Object next;
+            final boolean consumed;
             try {
-                next = buffer.take();
+                consumed = buffer.consume(this);
             } catch (InterruptedException e) {
                 return throwException(e);
             }
-            return processHasNext(next);
+            return terminal == null ? consumed : hasNextWhenTerminated();
         }
 
-        @Nullable
-        private T processNext() {
-            Object next = this.next;
-            this.next = null;
-            if (next == NULL_MASK) {
-                return null;
-            } else {
-                @SuppressWarnings("unchecked")
-                T t = (T) next;
-                return t;
-            }
+        private boolean hasNextWhenTerminated() {
+            assert terminal != null;
+            return terminal.cause() != null && (boolean) throwException(terminal.cause());
```

Comment:
`throwException` casting to a `boolean` doesn't look right, despite the return not being expected. Consider refactoring for clarity:

```java
Throwable cause = terminal.cause();
if (cause != null) {
  throwException(cause);
}
return false;
```